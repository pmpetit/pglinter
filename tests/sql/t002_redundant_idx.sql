-- Test for pglinter T002 rule: Tables with redundant indexes

\pset pager off

-- Create test tables with redundant indexes for T002 testing
CREATE TABLE IF NOT EXISTS customers_with_redundant_idx (
    customer_id INT PRIMARY KEY,
    first_name VARCHAR(100),
    last_name VARCHAR(100),
    email VARCHAR(255) UNIQUE,
    phone VARCHAR(20),
    created_at TIMESTAMP DEFAULT NOW(),
    updated_at TIMESTAMP DEFAULT NOW()
);

-- Create another table for more redundant index scenarios
CREATE TABLE IF NOT EXISTS products_with_redundant_idx (
    product_id SERIAL PRIMARY KEY,
    sku VARCHAR(50),
    name VARCHAR(255),
    category VARCHAR(100),
    price DECIMAL(10,2),
    stock_quantity INT,
    is_active BOOLEAN DEFAULT TRUE
);

-- Create a third table to test multiple tables with redundant indexes
CREATE TABLE IF NOT EXISTS orders_with_redundant_idx (
    order_id SERIAL PRIMARY KEY,
    customer_id INT,
    product_id INT,
    quantity INT,
    order_date DATE,
    status VARCHAR(50),
    total_amount DECIMAL(10,2)
);

-- Case 1: Exact duplicate indexes on customers table
CREATE INDEX idx_customers_email_1 ON customers_with_redundant_idx (email);
CREATE INDEX idx_customers_email_2 ON customers_with_redundant_idx (email);

-- Case 2: Same composite index created twice on customers table
CREATE INDEX idx_customers_name_1 ON customers_with_redundant_idx (first_name, last_name);
CREATE INDEX idx_customers_name_2 ON customers_with_redundant_idx (first_name, last_name);

-- Case 3: Redundant indexes on products table
CREATE INDEX idx_products_sku_1 ON products_with_redundant_idx (sku);
CREATE INDEX idx_products_sku_2 ON products_with_redundant_idx (sku);

-- Case 4: Complex composite index redundancy on products table
CREATE INDEX idx_products_category_active_1 ON products_with_redundant_idx (category, is_active);
CREATE INDEX idx_products_category_active_2 ON products_with_redundant_idx (category, is_active);

-- Case 5: Redundant indexes on orders table
CREATE INDEX idx_orders_customer_1 ON orders_with_redundant_idx (customer_id);
CREATE INDEX idx_orders_customer_2 ON orders_with_redundant_idx (customer_id);

-- Case 6: Different composite index redundancy on orders table
CREATE INDEX idx_orders_date_status_1 ON orders_with_redundant_idx (order_date, status);
CREATE INDEX idx_orders_date_status_2 ON orders_with_redundant_idx (order_date, status);

-- Case 7: Add some non-redundant indexes to ensure they don't trigger the rule
CREATE INDEX idx_customers_phone ON customers_with_redundant_idx (phone);
CREATE INDEX idx_products_price ON products_with_redundant_idx (price);
CREATE INDEX idx_orders_total ON orders_with_redundant_idx (total_amount);

CREATE EXTENSION IF NOT EXISTS pglinter;

-- First, disable all rules to isolate T002 testing
SELECT pglinter.disable_all_rules() AS all_rules_disabled;

-- Enable only T002 for focused testing
SELECT pglinter.enable_rule('T002') AS T002_enabled;

-- Test with file output
SELECT pglinter.perform_table_check('/tmp/pglinter_T002_results.sarif');

-- Test if file exists and show checksum
\! md5sum /tmp/pglinter_T002_results.sarif

-- Test with no output file (should output to prompt)
SELECT pglinter.perform_table_check();

-- Test rule management for T002
SELECT pglinter.explain_rule('T002');

-- Show that T002 is enabled
SELECT pglinter.is_rule_enabled('T002') AS T002_enabled;

-- Disable T002 temporarily and test
SELECT pglinter.disable_rule('T002') AS T002_disabled;
SELECT pglinter.perform_table_check();

-- Re-enable T002 and test again
SELECT pglinter.enable_rule('T002') AS T002_re_enabled;
SELECT pglinter.perform_table_check();

-- Clean up test tables
DROP TABLE IF EXISTS customers_with_redundant_idx CASCADE;
DROP TABLE IF EXISTS products_with_redundant_idx CASCADE;
DROP TABLE IF EXISTS orders_with_redundant_idx CASCADE;

