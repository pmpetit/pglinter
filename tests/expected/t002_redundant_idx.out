-- Test for pglinter T002 rule: Tables with redundant indexes
CREATE EXTENSION pglinter;
\pset pager off
-- First, disable all rules to isolate B001 testing
SELECT pglinter.disable_all_rules() AS all_rules_disabled;
NOTICE:  üî¥ Disabled 23 rule(s)
 all_rules_disabled 
--------------------
                 23
(1 row)

-- Enable only B001 for focused testing
SELECT pglinter.enable_rule('T002') AS b001_enabled;
NOTICE:  ‚úÖ Rule T002 has been enabled
 b001_enabled 
--------------
 t
(1 row)

-- Create test tables with redundant indexes for T002 testing
CREATE TABLE customers_with_redundant_idx (
    customer_id INT PRIMARY KEY,
    first_name VARCHAR(100),
    last_name VARCHAR(100),
    email VARCHAR(255) UNIQUE,
    phone VARCHAR(20),
    created_at TIMESTAMP DEFAULT NOW(),
    updated_at TIMESTAMP DEFAULT NOW()
);
-- Create another table for more redundant index scenarios
CREATE TABLE products_with_redundant_idx (
    product_id SERIAL PRIMARY KEY,
    sku VARCHAR(50),
    name VARCHAR(255),
    category VARCHAR(100),
    price DECIMAL(10, 2),
    stock_quantity INT,
    is_active BOOLEAN DEFAULT TRUE
);
-- Create a third table to test multiple tables with redundant indexes
CREATE TABLE orders_with_redundant_idx (
    order_id SERIAL PRIMARY KEY,
    customer_id INT,
    product_id INT,
    quantity INT,
    order_date DATE,
    status VARCHAR(50),
    total_amount DECIMAL(10, 2)
);
-- Case 1: Exact duplicate indexes on customers table
CREATE INDEX idx_customers_email_1 ON customers_with_redundant_idx (email);
CREATE INDEX idx_customers_email_2 ON customers_with_redundant_idx (email);
-- Case 2: Same composite index created twice on customers table
CREATE INDEX idx_customers_name_1 ON customers_with_redundant_idx (
    first_name, last_name
);
CREATE INDEX idx_customers_name_2 ON customers_with_redundant_idx (
    first_name, last_name
);
-- Case 3: Redundant indexes on products table
CREATE INDEX idx_products_sku_1 ON products_with_redundant_idx (sku);
CREATE INDEX idx_products_sku_2 ON products_with_redundant_idx (sku);
-- Case 4: Complex composite index redundancy on products table
CREATE INDEX idx_products_category_active_1 ON products_with_redundant_idx (
    category, is_active
);
CREATE INDEX idx_products_category_active_2 ON products_with_redundant_idx (
    category, is_active
);
-- Case 5: Redundant indexes on orders table
CREATE INDEX idx_orders_customer_1 ON orders_with_redundant_idx (customer_id);
CREATE INDEX idx_orders_customer_2 ON orders_with_redundant_idx (customer_id);
CREATE INDEX idx_orders_customer_3 ON orders_with_redundant_idx (customer_id, product_id);
CREATE INDEX idx_orders_customer_4 ON orders_with_redundant_idx (customer_id, product_id, order_date);
CREATE INDEX idx_orders_customer_5 ON orders_with_redundant_idx (customer_id, product_id, order_date, status);
-- Case 6: Different composite index redundancy on orders table
CREATE INDEX idx_orders_date_status_1 ON orders_with_redundant_idx (
    order_date, status
);
CREATE INDEX idx_orders_date_status_2 ON orders_with_redundant_idx (
    order_date, status
);
-- Case 7: Add some non-redundant indexes to ensure they don't trigger the rule
CREATE INDEX idx_customers_phone ON customers_with_redundant_idx (phone);
CREATE INDEX idx_products_price ON products_with_redundant_idx (price);
CREATE INDEX idx_orders_total ON orders_with_redundant_idx (total_amount);
-- First, disable all rules to isolate T002 testing
SELECT pglinter.disable_all_rules() AS all_rules_disabled;
NOTICE:  üî¥ Disabled 1 rule(s)
 all_rules_disabled 
--------------------
                  1
(1 row)

-- Enable only T002 for focused testing
SELECT pglinter.enable_rule('T002') AS t002_enabled;
NOTICE:  ‚úÖ Rule T002 has been enabled
 t002_enabled 
--------------
 t
(1 row)

-- Test with file output
SELECT pglinter.perform_table_check('/tmp/pglinter_T002_results.sarif');
 perform_table_check 
---------------------
 t
(1 row)

-- Test if file exists and show checksum
\! md5sum /tmp/pglinter_T002_results.sarif
e087b16a25e3c5367d95c44cecba0a62  /tmp/pglinter_T002_results.sarif
-- Test with no output file (should output to prompt)
SELECT pglinter.perform_table_check();
NOTICE:  üîç pglinter found 1 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [T002] WARNING: TABLE duplicated index 25 : 
public.customers_with_redundant_idx idx customers_with_redundant_idx_email_key columns (email) is a subset of idx idx_customers_email_1 columns (email)
public.customers_with_redundant_idx idx customers_with_redundant_idx_email_key columns (email) is a subset of idx idx_customers_email_2 columns (email)
public.customers_with_redundant_idx idx idx_customers_email_1 columns (email) is a subset of idx customers_with_redundant_idx_email_key columns (email)
public.customers_with_redundant_idx idx idx_customers_email_1 columns (email) is a subset of idx idx_customers_email_2 columns (email)
public.customers_with_redundant_idx idx idx_customers_email_2 columns (email) is a subset of idx customers_with_redundant_idx_email_key columns (email)
public.customers_with_redundant_idx idx idx_customers_email_2 columns (email) is a subset of idx idx_customers_email_1 columns (email)
public.customers_with_redundant_idx idx idx_customers_name_1 columns (first_name,last_name) is a subset of idx idx_customers_name_2 columns (first_name,last_name)
public.customers_with_redundant_idx idx idx_customers_name_2 columns (first_name,last_name) is a subset of idx idx_customers_name_1 columns (first_name,last_name)
public.orders_with_redundant_idx idx idx_orders_customer_1 columns (customer_id) is a subset of idx idx_orders_customer_2 columns (customer_id)
public.orders_with_redundant_idx idx idx_orders_customer_1 columns (customer_id) is a subset of idx idx_orders_customer_3 columns (customer_id,product_id)
public.orders_with_redundant_idx idx idx_orders_customer_1 columns (customer_id) is a subset of idx idx_orders_customer_4 columns (customer_id,product_id,order_date)
public.orders_with_redundant_idx idx idx_orders_customer_1 columns (customer_id) is a subset of idx idx_orders_customer_5 columns (customer_id,product_id,order_date,status)
public.orders_with_redundant_idx idx idx_orders_customer_2 columns (customer_id) is a subset of idx idx_orders_customer_1 columns (customer_id)
public.orders_with_redundant_idx idx idx_orders_customer_2 columns (customer_id) is a subset of idx idx_orders_customer_3 columns (customer_id,product_id)
public.orders_with_redundant_idx idx idx_orders_customer_2 columns (customer_id) is a subset of idx idx_orders_customer_4 columns (customer_id,product_id,order_date)
public.orders_with_redundant_idx idx idx_orders_customer_2 columns (customer_id) is a subset of idx idx_orders_customer_5 columns (customer_id,product_id,order_date,status)
public.orders_with_redundant_idx idx idx_orders_customer_3 columns (customer_id,product_id) is a subset of idx idx_orders_customer_4 columns (customer_id,product_id,order_date)
public.orders_with_redundant_idx idx idx_orders_customer_3 columns (customer_id,product_id) is a subset of idx idx_orders_customer_5 columns (customer_id,product_id,order_date,status)
public.orders_with_redundant_idx idx idx_orders_customer_4 columns (customer_id,product_id,order_date) is a subset of idx idx_orders_customer_5 columns (customer_id,product_id,order_date,status)
public.orders_with_redundant_idx idx idx_orders_date_status_1 columns (order_date,status) is a subset of idx idx_orders_date_status_2 columns (order_date,status)
public.orders_with_redundant_idx idx idx_orders_date_status_2 columns (order_date,status) is a subset of idx idx_orders_date_status_1 columns (order_date,status)
public.products_with_redundant_idx idx idx_products_category_active_1 columns (category,is_active) is a subset of idx idx_products_category_active_2 columns (category,is_active)
public.products_with_redundant_idx idx idx_products_category_active_2 columns (category,is_active) is a subset of idx idx_products_category_active_1 columns (category,is_active)
public.products_with_redundant_idx idx idx_products_sku_1 columns (sku) is a subset of idx idx_products_sku_2 columns (sku)
public.products_with_redundant_idx idx idx_products_sku_2 columns (sku) is a subset of idx idx_products_sku_1 columns (sku) 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 1 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
 perform_table_check 
---------------------
 t
(1 row)

-- Test rule management for T002
SELECT pglinter.explain_rule('T002');
NOTICE:  üìñ Rule Explanation for T002
============================================================

üéØ Rule Name: TableWithRedundantIndex
üìã Scope: TABLE

üìù Description:
table with duplicated index.

‚ö†Ô∏è  Message Template:
duplicated index

üîß How to Fix:
   1. remove duplicated index
   2. check for constraints that can create indexes.
============================================================
 explain_rule 
--------------
 t
(1 row)

-- Show that T002 is enabled
SELECT pglinter.is_rule_enabled('T002') AS t002_enabled;
 t002_enabled 
--------------
 t
(1 row)

-- Disable T002 temporarily and test
SELECT pglinter.disable_rule('T002') AS t002_disabled;
NOTICE:  üî¥ Rule T002 has been disabled
 t002_disabled 
---------------
 t
(1 row)

SELECT pglinter.perform_table_check();
NOTICE:  ‚úÖ No issues found - database schema looks good!
 perform_table_check 
---------------------
 t
(1 row)

-- Re-enable T002 and test again
SELECT pglinter.enable_rule('T002') AS t002_re_enabled;
NOTICE:  ‚úÖ Rule T002 has been enabled
 t002_re_enabled 
-----------------
 t
(1 row)

SELECT pglinter.perform_table_check();
NOTICE:  üîç pglinter found 1 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ö†Ô∏è  [T002] WARNING: TABLE duplicated index 25 : 
public.customers_with_redundant_idx idx customers_with_redundant_idx_email_key columns (email) is a subset of idx idx_customers_email_1 columns (email)
public.customers_with_redundant_idx idx customers_with_redundant_idx_email_key columns (email) is a subset of idx idx_customers_email_2 columns (email)
public.customers_with_redundant_idx idx idx_customers_email_1 columns (email) is a subset of idx customers_with_redundant_idx_email_key columns (email)
public.customers_with_redundant_idx idx idx_customers_email_1 columns (email) is a subset of idx idx_customers_email_2 columns (email)
public.customers_with_redundant_idx idx idx_customers_email_2 columns (email) is a subset of idx customers_with_redundant_idx_email_key columns (email)
public.customers_with_redundant_idx idx idx_customers_email_2 columns (email) is a subset of idx idx_customers_email_1 columns (email)
public.customers_with_redundant_idx idx idx_customers_name_1 columns (first_name,last_name) is a subset of idx idx_customers_name_2 columns (first_name,last_name)
public.customers_with_redundant_idx idx idx_customers_name_2 columns (first_name,last_name) is a subset of idx idx_customers_name_1 columns (first_name,last_name)
public.orders_with_redundant_idx idx idx_orders_customer_1 columns (customer_id) is a subset of idx idx_orders_customer_2 columns (customer_id)
public.orders_with_redundant_idx idx idx_orders_customer_1 columns (customer_id) is a subset of idx idx_orders_customer_3 columns (customer_id,product_id)
public.orders_with_redundant_idx idx idx_orders_customer_1 columns (customer_id) is a subset of idx idx_orders_customer_4 columns (customer_id,product_id,order_date)
public.orders_with_redundant_idx idx idx_orders_customer_1 columns (customer_id) is a subset of idx idx_orders_customer_5 columns (customer_id,product_id,order_date,status)
public.orders_with_redundant_idx idx idx_orders_customer_2 columns (customer_id) is a subset of idx idx_orders_customer_1 columns (customer_id)
public.orders_with_redundant_idx idx idx_orders_customer_2 columns (customer_id) is a subset of idx idx_orders_customer_3 columns (customer_id,product_id)
public.orders_with_redundant_idx idx idx_orders_customer_2 columns (customer_id) is a subset of idx idx_orders_customer_4 columns (customer_id,product_id,order_date)
public.orders_with_redundant_idx idx idx_orders_customer_2 columns (customer_id) is a subset of idx idx_orders_customer_5 columns (customer_id,product_id,order_date,status)
public.orders_with_redundant_idx idx idx_orders_customer_3 columns (customer_id,product_id) is a subset of idx idx_orders_customer_4 columns (customer_id,product_id,order_date)
public.orders_with_redundant_idx idx idx_orders_customer_3 columns (customer_id,product_id) is a subset of idx idx_orders_customer_5 columns (customer_id,product_id,order_date,status)
public.orders_with_redundant_idx idx idx_orders_customer_4 columns (customer_id,product_id,order_date) is a subset of idx idx_orders_customer_5 columns (customer_id,product_id,order_date,status)
public.orders_with_redundant_idx idx idx_orders_date_status_1 columns (order_date,status) is a subset of idx idx_orders_date_status_2 columns (order_date,status)
public.orders_with_redundant_idx idx idx_orders_date_status_2 columns (order_date,status) is a subset of idx idx_orders_date_status_1 columns (order_date,status)
public.products_with_redundant_idx idx idx_products_category_active_1 columns (category,is_active) is a subset of idx idx_products_category_active_2 columns (category,is_active)
public.products_with_redundant_idx idx idx_products_category_active_2 columns (category,is_active) is a subset of idx idx_products_category_active_1 columns (category,is_active)
public.products_with_redundant_idx idx idx_products_sku_1 columns (sku) is a subset of idx idx_products_sku_2 columns (sku)
public.products_with_redundant_idx idx idx_products_sku_2 columns (sku) is a subset of idx idx_products_sku_1 columns (sku) 

NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 1 warning(s), 0 info
NOTICE:  üü° Some warnings found - consider reviewing for optimization
 perform_table_check 
---------------------
 t
(1 row)

DROP TABLE orders_with_redundant_idx CASCADE;
DROP TABLE products_with_redundant_idx CASCADE;
DROP TABLE customers_with_redundant_idx CASCADE;
DROP EXTENSION pglinter CASCADE;
